### PR TITLE
fix cluster startup race

### LINKED ISSUE #5184: Can't start up Akka.Cluster when using custom task inlining dispatcher
**Version Information**
akka 1.4.22
akka, cluster

**Describe the bug**
After update to akka 1.4.22 from 1.4.21 cluster startup fails

Error: ValueFactory attempted to access the Value property of this instance.
System.InvalidOperationException: ValueFactory attempted to access the Value property of this instance.

```
...
[07:28:04 INF] Starting remoting
[07:28:04 DBG] Starting prune timer for endpoint manager...
Loaded '/app/bin/Debug/net5.0/DotNetty.Buffers.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Loaded '/app/bin/Debug/net5.0/DotNetty.Transport.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Loaded '/app/bin/Debug/net5.0/DotNetty.Common.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Loaded '/usr/share/dotnet/shared/Microsoft.NETCore.App/5.0.8/System.Reflection.TypeExtensions.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Loaded '/usr/share/dotnet/shared/Microsoft.NETCore.App/5.0.8/System.Reflection.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Loaded '/usr/share/dotnet/shared/Microsoft.NETCore.App/5.0.8/System.Reflection.Extensions.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
[07:28:04 INF] Remoting started; listening on addresses : [akka.tcp://pnet@pnetcore-01.cluster:2552]
[07:28:04 INF] Remoting now listens on addresses: [akka.tcp://pnet@pnetcore-01.cluster:2552]
[07:28:04 INF] Cluster Node [akka.tcp://pnet@pnetcore-01.cluster:2552] - Starting up...
[07:28:05 ERR] ValueFactory attempted to access the Value property of this instance.
System.InvalidOperationException: ValueFactory attempted to access the Value property of this instance.
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
--- End of stack trace from previous location ---
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.get_Value()
   at Akka.Actor.Internal.ActorSystemImpl.TryGetExtension[T](T& extension)
   at Akka.Actor.Internal.ActorSystemImpl.GetExtension[T]()
   at Akka.Actor.ActorSystemWithExtensions.WithExtension[T,TI](ActorSystem system)
   at Akka.Cluster.Cluster.Get(ActorSystem system)
   at Akka.Cluster.ClusterCoreSupervisor.<SupervisorStrategy>b__4_0(Exception e)
   at Akka.Actor.LocalOnlyDecider.Decide(Exception cause)
   at Akka.Actor.OneForOneStrategy.Handle(IActorRef child, Exception exception)
   at Akka.Actor.SupervisorStrategy.HandleFailure(ActorCell actorCell, IActorRef child, Exception cause, ChildRestartStats stats, IReadOnlyCollection`1 children)
   at Akka.Actor.ActorCell.HandleFailed(Failed f)
   at Akka.Actor.ActorCell.SysMsgInvokeAll(EarliestFirstSystemMessageList messages, Int32 currentState)
[07:28:05 ERR] ValueFactory attempted to access the Value property of this instance.
[akka://pnet/system/cluster/core#653436502]: Akka.Actor.PreRestartException: Exception pre restart (System.InvalidOperationException)
 ---> System.InvalidOperationException: ValueFactory attempted to access the Value property of this instance.
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
--- End of stack trace from previous location ---
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.get_Value()
   at Akka.Actor.Internal.ActorSystemImpl.TryGetExtension[T](T& extension)
   at Akka.Actor.Internal.ActorSystemImpl.GetExtension[T]()
   at Akka.Actor.ActorSystemWithExtensions.WithExtension[T,TI](ActorSystem system)
   at Akka.Cluster.Cluster.Get(ActorSystem system)
   at Akka.Cluster.ClusterCoreSupervisor.PostStop()
   at Akka.Actor.ActorBase.PreRestart(Exception reason, Object message)
   at Akka.Actor.ActorBase.AroundPreRestart(Exception cause, Object message)
   at Akka.Actor.ActorCell.FaultRecreate(Exception cause)
   --- End of inner exception stack trace ---
Loaded '/app/bin/Debug/net5.0/DotNetty.Handlers.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Loaded '/app/bin/Debug/net5.0/DotNetty.Codecs.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Loaded '/usr/share/dotnet/shared/Microsoft.NETCore.App/5.0.8/System.Runtime.Intrinsics.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
[07:28:18 DBG] Associated [akka.tcp://pnet@pnetcore-01.cluster:2552] <- akka.tcp://pnet@pnetdata-01.cluster:2552
Loaded '/app/bin/Debug/net5.0/Akka.Persistence.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
...
[07:28:25 ERR] Failed to startup Cluster. You can try to increase 'akka.actor.creation-timeout'.
Akka.Actor.AskTimeoutException: Timeout after 00:00:20 seconds
   at Akka.Cluster.Cluster.GetClusterCoreRef()
[07:28:25 INF] Cluster Node [akka.tcp://pnet@pnetcore-01.cluster:2552] - Shutting down...
[07:28:25 INF] Cluster Node [akka.tcp://pnet@pnetcore-01.cluster:2552] - Successfully shut down
[07:28:25 INF] Cluster Node [akka.tcp://pnet@pnetcore-01.cluster:2552] - Started up successfully
...

```

### PR DESCRIPTION
**Pull request by Zetanova**
There is a race condition in ClusterDomainEventPublisher and ClusterCoreDaemon both used Cluster.Get(system) to get the cluster instance. If the ActorCells are activated before the Cluster constructer gets the answer in  _clusterCore = GetClusterCoreRef().Result it will crash.

close #5184

### CODE REVIEW

**Comment by Aaronontheweb** [comment]
This definitely introduced some bugs - we might not be able to safely do this. The the blocking wait in the constructor was intentional.

**Comment by Aaronontheweb** [comment]
> This definitely introduced some bugs - we might not be able to safely do this. The the blocking wait in the constructor was intentional.

Not to say that it's impossible to fix, but it may not be so straightforward

**Comment by Zetanova** [comment]
I didn't remove the lock.

The PR passes the Cluster instance to the ClusterCoreDaemon directly
To remove Cluster.Get(system) from the constructor.

**Comment by Zetanova** [comment]
I only tested it with RemotePingPong and it worked.

**Comment by Aaronontheweb** [comment]
> I didn't remove the lock.
> 
> The PR passes the Cluster instance to the ClusterCoreDaemon directly
> To remove Cluster.Get(system) from the constructor.

Got it - I misunderstood. I'll take a closer look at this today.

**Comment by to11mtm** [comment]
@Zetanova I think you need to fix this line in ClusterDomainEventPublisherSpec.cs to get tests to pass:

```
//in public ClusterDomainEventPublisherSpec() : base(Config)
_publisher = Sys.ActorOf(Props.Create<ClusterDomainEventPublisher>()); //~line 66
```

`Props.Create` without a functor tries to look for a default constructor; due to your changes there is none here.

**Comment by Zetanova** [comment]
@to11mtm  thx, my misstake.
Why the multinode project now failed to build?

**Comment by Zetanova** [comment]
@to11mtm thx

Why `DDataStressTest.csproj` failed to build at `.net core unit test (linux)`  ?

**Comment by to11mtm** [comment]
> 
> 
> @to11mtm thx

Happy to help!

> 
> Why `DDataStressTest.csproj` failed to build at `.net core unit test (linux)` ?

Not entirely sure on that one, although I know that @Aaronontheweb has been working through some challenges on LMDB native dlls, might be related so tagging him.

**Comment by Aaronontheweb** [comment]
@Zetanova the DDataStressTest is a project we can possibly remove - we used it to help reproduce some data stemming from https://github.com/akkadotnet/akka.net/issues/5022

File: src/core/Akka.Cluster/ClusterDaemon.cs [hunk 1]
```diff
@@ -583,7 +583,7 @@ public override int GetHashCode()
         /// <summary>
         /// Gets a reference to the cluster core daemon.
         /// </summary>
-        internal class GetClusterCoreRef
+        internal class GetClusterCoreRef: INoSerializationVerificationNeeded
```
Comment (Aaronontheweb) [inline src/core/Akka.Cluster/ClusterDaemon.cs:586]: Good call

**Review by Aaronontheweb — APPROVED** [review]
LGTM

**Pull request merged**